platform/os_windows.go

```diff
@@ -257,40 +246,79 @@ func (p *execClient) ExecutePowershellCommandWithContext(ctx context.Context, co
-		log.Printf(errMsg)
-		return errors.Errorf(errMsg)
+		if errors.Is(err, registry.ErrNotExist) {
+			return nil
+		}
+		return errors.Wrap(err, "could not open registry key")
 	}
-	if strings.EqualFold(exists, "false") {
-		log.Printf("hns state path does not exist, skip setting SdnRemoteArpMacAddress")
-		return nil
+	defer k.Close()
+	// set the key value
+	if err = setSDNRemoteARPMACAddress(k); err != nil {
+		return errors.Wrap(err, "could not set registry key")
 	}
-	if sdnRemoteArpMacAddressSet == false {
-		result, err := execClient.ExecutePowershellCommand(GetSdnRemoteArpMacAddressCommand)
-		if err != nil {
-			return err
-		}
-
-		// Set the reg key if not already set or has incorrect value
-		if result != SDNRemoteArpMacAddress {
-			if _, err = execClient.ExecutePowershellCommand(SetSdnRemoteArpMacAddressCommand); err != nil {
-				log.Printf("Failed to set SDNRemoteArpMacAddress due to error %s", err.Error())
-				return err
-			}
+	// connect to the service manager
+	m, err := mgr.Connect()
+	if err != nil {
+		return errors.Wrap(err, "could not connect to service manager")
+	}
+	defer m.Disconnect() //nolint:errcheck // ignore error
+	// open the HNS service
+	service, err := m.OpenService("hns")
+	if err != nil {
+		return errors.Wrap(err, "could not access service")
+	}
+	defer service.Close()
+	if err := restartService(ctx, service); err != nil {
+		return errors.Wrap(err, "could not restart service")
+	}
+	return nil
+}
 
-			log.Printf("[Azure CNS] SDNRemoteArpMacAddress regKey set successfully. Restarting hns service.")
-			if _, err := execClient.ExecutePowershellCommand(RestartHnsServiceCommand); err != nil {
-				log.Printf("Failed to Restart HNS Service due to error %s", err.Error())
-				return err
-			}
-		}
+type key interface {
+	SetStringValue(name, value string) error
+}
 
-		sdnRemoteArpMacAddressSet = true
+func setSDNRemoteARPMACAddress(k key) error {
+	// Set the reg key
+	// was "Set-ItemProperty -Path HKLM:\\SYSTEM\\CurrentControlSet\\Services\\hns\\State -Name SDNRemoteArpMacAddress -Value \"12-34-56-78-9a-bc\""
+	if err := k.SetStringValue("SDNRemoteArpMacAddress", SDNRemoteArpMacAddress); err != nil {
+		return errors.Wrap(err, "could not set registry key")
 	}
+	log.Printf("[Azure CNS] SDNRemoteArpMacAddress regKey set successfully. Restarting hns service.")
+	return nil
+}
 
+func restartService(ctx context.Context, s *mgr.Service) error {
+	// Stop the service
+	_, err := s.Control(svc.Stop)
+	if err != nil {
+		return errors.Wrap(err, "could not stop service")
+	}
+	// Wait for the service to stop
+	ticker := time.NewTicker(500 * time.Millisecond) //nolint:gomnd // 500ms
+	defer ticker.Stop()
+	for { // hacky cancellable do-while
+		status, err := s.Query()
+		if err != nil {
+			return errors.Wrap(err, "could not query service status")
+		}
+		if status.State == svc.Stopped {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return errors.New("context cancelled")
+		case <-ticker.C:
+		}
+	}
```

Comment:
You have not started the ticker. Should we add the ticker to the loop to start the ticker and make the loop finite
```suggestion
	for range ticker.C { // hacky cancellable do-while
		status, err := s.Query()
		if err != nil {
			return errors.Wrap(err, "could not query service status")
		}
		if status.State == svc.Stopped {
			break
		}
		select {
		case <-ctx.Done():
			return errors.New("context cancelled")
		}
	}
```

Reply:
I don't know what you mean by "have not started the ticker". [Tickers](https://gobyexample.com/tickers) are immediately ready to use on initialization.

NACK on your suggestion. You have changed the semantics from `do-while`, to simply `while`. I want to test the condition first, then wait, then repeat. As a `while`, we wait first, then test the condition, then repeat.

Reply:
Good to NACK the suggestion the Copilot thread I read suggested suggested the following:
```
package main

import (
    "fmt"
    "time"
)

func main() {
    fmt.Println("Start Ticker Example")
    // Define an interval and create a new Ticker
    interval := time.Duration(2) * time.Second
    ticker := time.NewTicker(interval)
    defer ticker.Stop()

    // Start the ticker by constructing a loop
    i := 0
    for range ticker.C {
        i++
        countFuncCall(i)
        if i >= 5 {
            break
        }
    }
}

// Define the function to be called
func countFuncCall(i int) {
    fmt.Println("Function is called:", i, "times")
}
```
So I assumed you have to start the Ticker. Thanks for the clarification. Resolving this.